File: scripts/erp-local-print-agent.js
/**
 * KC ERP — local Windows USB label print agent.
 * Run on the shop PC: node scripts/erp-local-print-agent.js
 * Sends raw TSPL to the TSC printer installed as USB001 (Windows spooler name).
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { execFile } = require('child_process');
const { promisify } = require('util');

const execFileAsync = promisify(execFile);
const PORT = Number(process.env.ERP_PRINT_AGENT_PORT || 17888);
const HOST = '127.0.0.1';
const PS1 = path.join(__dirname, 'windows-raw-print.ps1');

async function printRaw(printerName, tspl) {
    const tmp = path.join(os.tmpdir(), `kc-erp-label-${Date.now()}-${Math.random().toString(36).slice(2, 8)}.prn`);
    fs.writeFileSync(tmp, String(tspl || ''), 'utf8');
    try {
        await execFileAsync(
            'powershell.exe',
            ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', PS1, '-PrinterName', printerName, '-FilePath', tmp],
            { timeout: 45000, windowsHide: true },
        );
    } finally {
        try {
            fs.unlinkSync(tmp);
        } catch {
            /* ignore */
        }
    }
}

async function printRawBinary(printerName, escPosBase64) {
    const tmp = path.join(os.tmpdir(), `kc-erp-receipt-${Date.now()}-${Math.random().toString(36).slice(2, 8)}.bin`);
    fs.writeFileSync(tmp, Buffer.from(String(escPosBase64 || ''), 'base64'));
    try {
        await execFileAsync(
            'powershell.exe',
            ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-File', PS1, '-PrinterName', printerName, '-FilePath', tmp],
            { timeout: 45000, windowsHide: true },
        );
    } finally {
        try {
            fs.unlinkSync(tmp);
        } catch {
            /* ignore */
        }
    }
}

function readBody(req) {
    return new Promise((resolve, reject) => {
        let body = '';
        req.on('data', (chunk) => {
            body += chunk;
            if (body.length > 2_000_000) reject(new Error('Payload too large'));
        });
        req.on('end', () => resolve(body));
        req.on('error', reject);
    });
}

const server = http.createServer(async (req, res) => {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        res.writeHead(204);
        res.end();
        return;
    }

    if (req.method === 'GET' && req.url === '/health') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ ok: true, service: 'kc-erp-local-print', port: PORT }));
        return;
    }

    if (req.method === 'GET' && req.url === '/printers') {
        try {
            const { stdout } = await execFileAsync(
                'powershell.exe',
                [
                    '-NoProfile',
                    '-Command',
                    'Get-Printer | Select-Object -ExpandProperty Name | ConvertTo-Json -Compress',
                ],
                { timeout: 15000, windowsHide: true },
            );
            let names = [];
            try {
                const parsed = JSON.parse(stdout || '[]');
                names = Array.isArray(parsed) ? parsed : parsed ? [parsed] : [];
            } catch {
                names = [];
            }
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: true, printers: names }));
        } catch (e) {
            res.writeHead(500, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: false, error: e.message || 'Could not list printers' }));
        }
        return;
    }

    if (req.method === 'POST' && req.url === '/print') {
        try {
            const body = await readBody(req);
            const payload = JSON.parse(body || '{}');
            const printerName = String(payload.printerName || 'TSC TTP-244 Pro').trim();

            if (payload.escPosBase64) {
                await printRawBinary(printerName, payload.escPosBase64);
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({ ok: true, count: 1, printerName, kind: 'receipt' }));
                return;
            }

            const list = Array.isArray(payload.tsplList)
                ? payload.tsplList
                : payload.tspl
                  ? [payload.tspl]
                  : [];
            if (!list.length) {
                res.writeHead(400, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({ ok: false, error: 'No TSPL or receipt data' }));
                return;
            }
            for (let i = 0; i < list.length; i += 1) {
                await printRaw(printerName, list[i]);
                if (i < list.length - 1) {
                    await new Promise((r) => setTimeout(r, 450));
                }
            }
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: true, count: list.length, printerName, kind: 'label' }));
        } catch (e) {
            res.writeHead(500, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: false, error: e.message || 'Print failed' }));
        }
        return;
    }

    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: false, error: 'Not found' }));
});

server.listen(PORT, HOST, () => {
    console.log(`KC ERP local print agent listening on http://${HOST}:${PORT}`);
    console.log('Keep this window open while printing labels from the ERP in Chrome.');
    console.log('Default printer name: TSC TTP-244 Pro (change in Hardware → USB settings).');
});

File: scripts/erp-local-print-agent.py
#!/usr/bin/env python3
"""
KC ERP — local Windows USB label print agent (Python).
Optional alternative if Python is installed. Usually use start-erp-print-agent.bat instead.
"""

from __future__ import annotations

import base64
import json
import os
import subprocess
import sys
import tempfile
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from typing import Any

PORT = int(os.environ.get("ERP_PRINT_AGENT_PORT", "17888"))
HOST = "127.0.0.1"
SCRIPT_DIR = Path(__file__).resolve().parent
RAW_PRINT_PS1 = SCRIPT_DIR / "windows-raw-print.ps1"


def print_raw_binary(printer_name: str, esc_pos_base64: str) -> None:
    suffix = os.urandom(4).hex()
    fd, tmp = tempfile.mkstemp(prefix=f"kc-erp-receipt-{suffix}-", suffix=".bin")
    os.close(fd)
    try:
        Path(tmp).write_bytes(base64.b64decode(str(esc_pos_base64 or "")))
        result = subprocess.run(
            [
                "powershell.exe",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-File",
                str(RAW_PRINT_PS1),
                "-PrinterName",
                printer_name,
                "-FilePath",
                tmp,
            ],
            capture_output=True,
            text=True,
            timeout=45,
            check=False,
        )
        if result.returncode != 0:
            detail = (result.stderr or result.stdout or "").strip()
            raise RuntimeError(detail or f"Receipt print failed for {printer_name}")
    finally:
        try:
            os.unlink(tmp)
        except OSError:
            pass


def print_raw(printer_name: str, tspl: str) -> None:
    suffix = os.urandom(4).hex()
    fd, tmp = tempfile.mkstemp(prefix=f"kc-erp-label-{suffix}-", suffix=".prn")
    os.close(fd)
    try:
        Path(tmp).write_text(str(tspl or ""), encoding="utf-8")
        result = subprocess.run(
            [
                "powershell.exe",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-File",
                str(RAW_PRINT_PS1),
                "-PrinterName",
                printer_name,
                "-FilePath",
                tmp,
            ],
            capture_output=True,
            text=True,
            timeout=45,
            check=False,
        )
        if result.returncode != 0:
            detail = (result.stderr or result.stdout or "").strip()
            raise RuntimeError(detail or f"Raw print failed for {printer_name}")
    finally:
        try:
            os.unlink(tmp)
        except OSError:
            pass


class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt: str, *args: Any) -> None:
        return

    def _cors(self) -> None:
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def _json(self, status: int, payload: dict[str, Any]) -> None:
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self._cors()
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self) -> None:
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_GET(self) -> None:
        if self.path.rstrip("/") == "/health":
            self._json(
                200,
                {
                    "ok": True,
                    "service": "kc-erp-local-print",
                    "port": PORT,
                    "runtime": "python",
                    "supportsReceipt": True,
                    "supportsLabels": True,
                },
            )
            return
        if self.path.rstrip("/") == "/printers":
            try:
                out = subprocess.check_output(
                    [
                        "powershell.exe",
                        "-NoProfile",
                        "-Command",
                        "Get-Printer | Select-Object -ExpandProperty Name | ConvertTo-Json -Compress",
                    ],
                    text=True,
                    timeout=15,
                )
                parsed = json.loads(out or "[]")
                names = parsed if isinstance(parsed, list) else [parsed] if parsed else []
                self._json(200, {"ok": True, "printers": names})
            except Exception as exc:
                self._json(500, {"ok": False, "error": str(exc)})
            return
        self._json(404, {"ok": False, "error": "Not found"})

    def do_POST(self) -> None:
        path = self.path.rstrip("/")
        if path == "/print-receipt":
            try:
                length = int(self.headers.get("Content-Length", "0"))
                if length > 2_000_000:
                    raise ValueError("Payload too large")
                raw = self.rfile.read(length).decode("utf-8") if length else "{}"
                payload = json.loads(raw or "{}")
                printer_name = str(payload.get("printerName") or "EPSON TM-m30III Receipt").strip()
                if not payload.get("escPosBase64"):
                    self._json(400, {"ok": False, "error": "No receipt data"})
                    return
                print_raw_binary(printer_name, str(payload["escPosBase64"]))
                self._json(200, {"ok": True, "count": 1, "printerName": printer_name, "kind": "receipt"})
            except Exception as exc:
                self._json(500, {"ok": False, "error": str(exc)})
            return
        if path != "/print":
            self._json(404, {"ok": False, "error": "Not found"})
            return
        try:
            length = int(self.headers.get("Content-Length", "0"))
            if length > 2_000_000:
                raise ValueError("Payload too large")
            raw = self.rfile.read(length).decode("utf-8") if length else "{}"
            payload = json.loads(raw or "{}")
            printer_name = str(payload.get("printerName") or "TSC TTP-244 Pro").strip()
            if payload.get("escPosBase64"):
                print_raw_binary(printer_name, str(payload["escPosBase64"]))
                self._json(200, {"ok": True, "count": 1, "printerName": printer_name, "kind": "receipt"})
                return
            tspl_list = payload.get("tsplList")
            if tspl_list is None and payload.get("tspl"):
                tspl_list = [payload["tspl"]]
            tspl_list = list(tspl_list or [])
            if not tspl_list:
                self._json(400, {"ok": False, "error": "No TSPL or receipt data"})
                return
            for i, item in enumerate(tspl_list):
                print_raw(printer_name, str(item))
                if i < len(tspl_list) - 1:
                    time.sleep(0.45)
            self._json(200, {"ok": True, "count": len(tspl_list), "printerName": printer_name})
        except Exception as exc:
            self._json(500, {"ok": False, "error": str(exc)})


def main() -> None:
    print("")
    print("========================================")
    print(" KC ERP Label Print Service (USB001)")
    print("========================================")
    print(f" Listening on http://{HOST}:{PORT}/")
    print(" Keep this window OPEN while printing from Chrome.")
    print(" Press Ctrl+C to stop.")
    print("")
    server = HTTPServer((HOST, PORT), Handler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()


if __name__ == "__main__":
    main()
